src/GitHub.Api/Tasks/ActionTask.cs

```diff
@@ -5,6 +5,53 @@
 
 namespace GitHub.Unity
 {
+    class TaskQueue : TaskBase
+    {
+        private TaskCompletionSource<bool> aggregateTask = new TaskCompletionSource<bool>();
+        private readonly List<ITask> queuedTasks = new List<ITask>();
+        private volatile bool isSuccessful = true;
+        private volatile Exception exception;
+        private int finishedTaskCount;
+
+        public TaskQueue() : base()
+        {
+            Initialize(aggregateTask.Task);
+        }
+
+        public ITask Queue(ITask task)
+        {
+            task.OnEnd += TaskFinished;
+            queuedTasks.Add(task);
+            return this;
+        }
+
+        protected override void Run()
+        {
+            foreach (var task in queuedTasks)
+                task.Start();
+            base.Run();
+        }
+
+        private void TaskFinished(ITask task, bool success, Exception ex)
+        {
+            var count = Interlocked.Increment(ref finishedTaskCount);
```

Comment:
interlocked increment followed by a code that is not thread safe gives a false promise of threading.

Declaring isSuccessful volatile is not enough to make this statement thread safe: isSuccessful &= success 

On top you have a race condition to set isSuccessful and exception prior to reading them below.

Reply:
Hmm, that message disappeared into outdated... reposting.

Super awesome that you're so responsive. Two out of three bugs closed already, however when is the next release? I'd like to present this to our teams on Friday, but I'd like the two closed bugs in before I do. About the race condition here I'm not super worried about it... the task completes and the icons will update correctly, that the dialog error message fails once in a thousand is hardly a blocker for me as long as TrySetException(null) doesn't blow up in our face. 
:thumbsup:

Reply:
to fix the threading issues just go this route:
```
            if (!success)
            {
                isSuccessful = false;
                exception = ex;
            }
            var count = Interlocked.Increment(ref finishedTaskCount);
```
There...  everything is thread safe, it is safe to set the volatile isSuccessful and exception. It's important to only increment after you've updated the internal state so that when count reaches queuedTasks.Count you're guaranteed that isSuccessful and exception has been set by all threads (if they needed changing that is).

Reply:
Hey @Joen-UnLogick thanks for testing out our stuff and liking it enough to share with your team. Chances are we will be pushing out another release today or tomorrow.

I hadn't addressed the threading issue yet because I haven't yet wrapped my head around it, if your fix covers the issue, please don't hesitate to send a pull request. I'd like to give you credit if you fixed the issue.

Reply:
@Joen-UnLogick  Yeah, I wrote that bit in a rush, gotta love threading.

> Super awesome that you're so responsive. Two out of three bugs closed already, however when is the next release?

I'm getting an rc3 ready over at the `releases/1.0.0rc3` branch, we can cut a build for you with all the fixes. Which bugs do you care more about?

Reply:
@shana Well #785 is the most important one, it is mostly artists and designers who will be locking stuff in Unity and the lack of auto refresh is a blocker. I took a look at your fix and determined I didn't know the internals of your system well enough to determine if the fix is good or not. The rest of the issues like this one isn't blocking in the same sense.